### ISSUE TITLE
Support for Kolmogorov-Arnold Networks (KANs)

### ISSUE THREAD

**Issue by anshika-anand** [pytorch/pytorch#125389]
### 🚀 The feature, motivation and pitch

Proposal for adding support for Kolmogorov-Arnold Networks (KANs) in PyTorch. KANs are a promising alternative to traditional Multi-Layer Perceptrons (MLPs) and offer potential advantages in terms of accuracy, scaling, and interpretability.

KANs are inspired by the Kolmogorov-Arnold representation theorem and have a fundamentally different structure compared to MLPs. Instead of having fixed activation functions on nodes (neurons) and linear weights, KANs have learnable activation functions on edges (weights), and every weight parameter is replaced by a univariate function parameterized as a spline.

### Additional context

[KAN: Kolmogorov–Arnold Networks](https://arxiv.org/pdf/2404.19756), The paper was submitted on 30th April,2024 by researchers from Massachusetts Institute of Technology, California Institute of Technology, Northeastern University, The NSF Institute for Artificial Intelligence and Fundamental Interactions.

**Comment by hvaara** [comment]
Ref https://github.com/KindXiaoming/pykan
cc @KindXiaoming

**Comment by jbschlosser** [comment]
Hey @anshika-anand, thanks for the request! To keep our maintenance burden manageable, PyTorch Core has fairly strict rules regarding adding a new Optimizer or Module function or feature: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision. 

Currently, KANs seem best suited to be utilized from an external library such as [pykan](https://github.com/KindXiaoming/pykan) linked above. In the future, we'll happily reconsider if this function increases in popularity to the point where it is expected within `torch.nn`. We can leave this issue open to gauge popularity among other researchers and users.

**Comment by rafa-br34** [comment]
I think having KANs built-in with `torch.nn` could aid in further AI development as it has shown promising results. Still, waiting a bit before adding it to PyTorch seems to be the proper move as it is an architecture that only recently has been properly defined and researched.

**Comment by hvaara** [comment]
For those coming here from Google, I'd like to leave https://github.com/mintisan/awesome-kan as a breadcrumb. It's a great entry point to learn more - papers, projects, tutorials, etc - and to find a suitable library when working on a problem with code.
